Fix: Product categories widget item count not always showing the correct number

roykho opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #25375

### How to test the changes in this Pull Request:

1. Add the Product Categories widget to your site. Enable show counts option.
2. Make sure to have a hand full of products that belong to a category that is listed in that widget.
3. For couple of the products, set the stock level to out of stock and save the product.
4. Go back to the frontend shop page and ensure the widget shows the product count correctly.
5. Go to your woocommerce->settings->products->inventory and set `Hide out of stock items from the catalog` to enabled.
6. Go back to the frontend shop page and ensure the widget shows the product count correctly. Should show less than in step 5 as now it suppose to hide the product from catalog if it is out of stock.
7. Disable `Hide out of stock items from the catalog` setting and save.
8. Go back to the frontend shop page and ensure the widget shows the product count correctly.
9. Test the added filter by using the code snippet `add_filter( 'woocommerce_product_recount_terms', '__return_false' );`.
10. Test steps 1 to 8 and observe that the count is no longer refreshed because the filter prevents the recount.

### Other information:

Simple PR to just flush the caches during the product save operation and product settings page.

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Fix - Product categories widget item count not always showing the correct number.
> Add - Filter `woocommerce_product_recount_terms` to allow/prevent recounting of product terms.

roykho committed:
Extract recount terms logic into own function closes #25375

roykho committed:
Apply recount term per product with filter to product data store

Konamiman commented:
Perhaps I'm doing something wrong, but what I'm seeing is that the counts in the widget are always refreshed (and showing the correct amount) on shop page reload when I change something (enable/disable the "hide out of stock products" option, or set a product in/out of stock), even if I use the filter to disable the behavior (and via debugging I have verified that the filter is indeed working as expected).

roykho commented:
@Konamiman - so without this PR, if you play around with the `out of stock` setting per product and even make the product `hidden` and reload the front page with the widget, you will see it will sometimes not refresh and you will see the old count.

roykho commented:
@Konamiman - for example ![](https://d.pr/i/Gq1WBF+)

Konamiman commented:
@roykho Sure, that works as it should. What I haven't been able to verify is the behavior of the filter: even if I set `woocommerce_product_update_recount_terms` to `return false`, the counters always get updated; I'm not sure if that's something bad.

roykho commented:
@Konamiman - when the filter is set to return false, after changing the settings, it will not update for me.

roykho commented:
@Konamiman - Just wanted to add more context. You actually reviewed another PR here https://github.com/woocommerce/woocommerce/pull/28974 it is the same issue just different approach. So it seem you could reproduce the issue then?

Konamiman commented:
@roykho I've been debugging in more detail. I have found that when changing the stock status of a product, and despite the filter being in place and returning `false`, the `_wc_term_recount` function is called anyway. The (reverse) stack trace is as follows:

* `_wc_term_recount` is called via `$taxonomy->update_count_callback` inside `wp_update_term_count_now`
* Which is called by `wp_update_term_count`
* Which is called by `_update_term_count_on_transition_post_status`
* Which is called via `do_action('transition_post_status')` inside `wp_transition_post_status`
* Which is called by `wp_insert_post`
* Which is called by `wp_update_post`
* Which is called by  `edit_post`
* Which is called from inside `wp-admin/post.php`

The weird thing is that this is not happening to you!

> So it seem you could reproduce the issue then?

The issue is with the behavior of the filter, which didn't exist in that PR.

roykho commented:
@Konamiman - so instead of the stock status have you tried enabling/disabling this setting ![](https://d.pr/i/Gat9b5+)

And also try this ![](https://d.pr/i/dKO9dK+)

Konamiman commented:
@roykho Yes, I've tried both and in both cases the counter gets updated as well.

* When setting the product as "Hidden", the stack trace is the same as when changing the stock status of the product.
* When changing the "Hide out of stock products" setting, `_wc_term_recount` is called by `wc_recount_all_terms`, which is called by `WC_Settings_Products::save`

Konamiman commented:
> Are you saying this PR works as should but only the filter is not working?

Yes, exactly that. My guess is that the filter just needs to be applied in a couple more of places (the ones I found while debugging).

roykho committed:
Refactor and moved filter to _wc_recount_terms function

roykho commented:
@Konamiman - Ok thanks for confirming. I have moved the filter to `_wc_recount_terms` so it catches all cases (that I've found) that calls this function. Also because of this move, the filter name doesn't make sense any more so I renamed it to `add_filter( 'woocommerce_product_recount_terms', '__return_false' );` So when you test, please use this new filter name.

Konamiman commented:
Great! It all works as expected now.
